src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeConverter.cs

```diff
@@ -253,63 +231,70 @@ static string UnexpectedNodeType(RegexNode node)
                                 // Alternations are created by creating an Or of all of its children.
                                 // This Or needs to be "ordered" to achieve the same semantics as the backtracking engines.
                                 SymbolicRegexNode<BDD> or = _builder._nothing;
-                                // enumerate in reverse order through the child results
-                                for (int i = childResults.Length - 1; i >= 0; --i)
+
+                                // Enumerate in reverse order through the child results
+                                for (int i = childResults.Length - 1; i >= 0; i--)
                                 {
-                                    ConversionResult res = childResults[i];
-                                    // if res is a non-singleton list then it denotes a concatenation that must be constructed at this point
-                                    SymbolicRegexNode<BDD> elem = res._size == 1 ? res.FirstElement : _builder.CreateConcatRev(res.Enumerate(true), top.TryToMarkFixedLength);
+                                    DoublyLinkedList<SymbolicRegexNode<BDD>> childResult = childResults[i];
+
+                                    // If childResult is a non-singleton list, then it denotes a concatenation that must be constructed at this point.
+                                    SymbolicRegexNode<BDD> elem = childResult.Count == 1 ?
+                                        childResult.FirstElement :
+                                        _builder.CreateConcatAlreadyReversed(childResult.EnumerateLastToFirst(), popped.TryToMarkFixedLength);
                                     if (elem.IsNothing)
                                     {
                                         continue;
                                     }
 
-                                    or = SymbolicRegexNode<BDD>.OrderedOr(_builder, elem, or);
-
-                                    if (elem.IsAnyStar)
-                                        or = elem; // .* is the absorbing element
+                                    or = elem.IsAnyStar ?
+                                        elem : // .* is the absorbing element
+                                        SymbolicRegexNode<BDD>.OrderedOr(_builder, elem, or);
```

Comment:
I assume OrderedOr doesn't have side-effects, and so in the elem.IsAnyStar we can just skip the OrderedOr call.